subsys/bluetooth/host/hci_core.c

```diff
@@ -4919,6 +4931,155 @@ static void le_adv_ext_report(struct net_buf *buf)
 		net_buf_pull(buf, evt->length);
 	}
 }
+
+#if defined(CONFIG_BT_PER_ADV_SYNC)
+static struct bt_le_per_adv_sync *get_pending_per_adv_sync(void)
+{
+	for (int i = 0; i < ARRAY_SIZE(per_adv_sync_pool); i++) {
+		if (atomic_test_bit(per_adv_sync_pool[i].flags,
+				    BT_PER_ADV_SYNC_SYNCING)) {
+			return &per_adv_sync_pool[i];
+		}
+	}
+
+	return NULL;
+}
+
+static struct bt_le_per_adv_sync *get_per_adv_sync(uint16_t handle)
+{
+	for (int i = 0; i < ARRAY_SIZE(per_adv_sync_pool); i++) {
+		if (per_adv_sync_pool[i].handle == handle &&
+		    atomic_test_bit(per_adv_sync_pool[i].flags,
+				    BT_PER_ADV_SYNC_SYNCED)) {
+			return &per_adv_sync_pool[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void le_per_adv_report(struct net_buf *buf)
+{
+	struct bt_hci_evt_le_per_advertising_report *evt;
+	struct bt_le_per_adv_sync *per_adv_sync;
+	struct bt_le_per_adv_sync_recv_info info;
+
+	if (buf->len < sizeof(*evt)) {
+		BT_ERR("Unexpected end of buffer");
+		return;
+	}
+
+	evt = net_buf_pull_mem(buf, sizeof(*evt));
+
+	per_adv_sync = get_per_adv_sync(sys_le16_to_cpu(evt->handle));
+
+	if (!per_adv_sync) {
+		BT_ERR("Unknown handle 0x%04X for periodic advertising report",
+		       sys_le16_to_cpu(evt->handle));
+		return;
+	}
+
+	info.tx_power = evt->tx_power;
+	info.rssi = evt->rssi;
+	info.cte_type = evt->cte_type;
+	info.addr = &per_adv_sync->addr;
+
+	if (per_adv_sync->cb && per_adv_sync->cb->recv) {
+		per_adv_sync->cb->recv(per_adv_sync, &info, &buf->b);
+	}
+}
+
+static int per_adv_sync_terminate(uint16_t handle)
+{
+	struct bt_hci_cp_le_per_adv_terminate_sync *cp;
+	struct net_buf *buf;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_PER_ADV_TERMINATE_SYNC,
+				sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	(void)memset(cp, 0, sizeof(*cp));
+
+	cp->handle = sys_cpu_to_le16(handle);
+
+	return bt_hci_cmd_send_sync(BT_HCI_OP_LE_PER_ADV_TERMINATE_SYNC, buf,
+				    NULL);
+}
+
+static void le_per_adv_sync_established(struct net_buf *buf)
```

Comment:
How is this handled?

> When the HCI_LE_Periodic_Advertising_Create_Sync_Cancel command has
completed, the Controller sends an HCI_Command_Complete event to the
Host.
After the HCI_Command_Complete is sent and if the cancellation was
successful, the Controller sends an
HCI_LE_Periodic_Advertising_Sync_Established event to the Host with the
error code Operation Cancelled by Host (0x44).

Reply:
Generally it didn't do anything about any errors. I changed it to call the terminate callback in case that the HCI error code isn't success.

Reply:
I suppose this is similar to how le_create_connection_cancel works, see enh_conn_complete handling of BT_HCI_ERR_UNKNOWN_CONN_ID status.

In this case the application is notified with a "connected" callback with error code = 0x02.

Question is if we should do it similar for this event then. To call "synced" callback instead with an error code. It would be consistent, but not sure it is the best.

For the connected callback this is an error handling that many forget, I think it might have been better if disconnected callback would have been called instead.

Reply:
I agree with @joerchan . In my mind the disconnected/terminated callback makes more sense if a connection/sync isn't established due to an error. 

Although I'm not sure if it's better to change it here either; consistency is sometimes better. 

Unless someone requests a change (@cvinayak ?) to call the sync callback here instead, I propose to leave it as is. 

I did consider adding a `reason` field in the terminated callback (as it would of course be nice for the application to know why it failed), but when a sync has been established and is then terminated, there isn't a reason in the event. Of course we can just add our own reasons:
1) Sync failed
2) Locally terminated
3) Remote terminated

Reply:
![image](https://user-images.githubusercontent.com/6350656/89649873-c1cdd000-d8c1-11ea-96a7-aea43996ecff.png)

This is how it appears in the log for Cancelling Periodic Sync. By HCI Spec. the LE Periodic Advertising Terminate Sync command will return error 0xx42.
` If the periodic advertising train corresponding to the Sync_Handle parameter
does not exist, then the Controller shall return the error code Unknown
Advertising Identifier (0x42).`

Reply:
@Emil-Gydesen-Bose Sorry, I forgot an important detail regarding the create_conn_cancel, the callback is only called if the cancel was initiated by the host from the timer (CONFIG_BT_CREATE_CONN_TIMEOUT). If initiated by the API then the callback is not called.

Since there is no timeout internally in the host in this case, I suggest to not call any callback for this.
And also there shouldn't be any ERR or WRN in the log in something that is expected as Vinayak points out.

Reply:
@joerchan @cvinayak I removed the callback on error from the event handler. 

As far as I can tell, if the cancel command has the success status code, then the sync event will have the status 0x44. To me that sounds like we don't need to wait for the sync event until we declare the adv_sync struct deleted, which is done in the function that sends the cancel command (which waits for the command complete event). 

In other words, the sync established event handler simply returns if the sync create is cancelled, as all the cleanup is already handled elsewhere. 

@cvinayak The status code 0x42 will be in the command complete event, when is handled by `bt_le_per_adv_sync_delete`. 0x42 should never appear in the sync established event.